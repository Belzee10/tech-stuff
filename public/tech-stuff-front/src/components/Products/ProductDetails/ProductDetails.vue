<template>
  <v-row>
    <v-col cols="10" offset="1">
      <v-card class="product-details">
        <v-container class="py-0 mt-10">
          <v-row v-if="!product">
            <v-col>
              <span>Loading...</span>
            </v-col>
          </v-row>
          <v-row v-else justify="space-between">
            <v-col class="pa-0" cols="auto">
              <v-img
                height="400"
                width="400"
                :src="product.image"
                lazy-src="https://picsum.photos/id/11/10/6"
              ></v-img>
            </v-col>
            <v-col class="d-flex flex-column">
              <h4 class="display-1">{{ product.name }}</h4>
              <span class="subtitle-1 d-block mt-2">{{ product.brand }}</span>
              <span class="font-weight-bold d-block">${{ product.price }}</span>
              <p class="mt-10 font-weight-light">{{ product.description }}</p>
              <v-row>
                <v-col
                  cols="4"
                  offset="8"
                  class="d-flex justify-space-between mt-auto"
                >
                  <v-btn
                    v-if="product.inStock > 0"
                    small
                    depressed
                    class="btn-small"
                    color="secondary"
                    @click="removeFromCart"
                  >
                    <v-icon small>mdi-cart-minus</v-icon>
                  </v-btn>
                  <span v-if="product.inStock > 0">
                    {{ product.inStock }} in Cart
                  </span>
                  <v-btn
                    v-if="product.inStock > 0"
                    small
                    depressed
                    class="btn-small"
                    color="secondary"
                    @click="addToCart"
                  >
                    <v-icon small>mdi-cart-plus</v-icon>
                  </v-btn>
                  <v-btn
                    v-else
                    small
                    block
                    depressed
                    class="btn-full text-capitalize"
                    color="secondary"
                    @click="addToCart"
                  >
                    <v-icon small class="mr-1">mdi-cart-plus</v-icon>
                    Add to Cart
                  </v-btn>
                </v-col>
              </v-row>
            </v-col>
          </v-row>
        </v-container>
      </v-card>
    </v-col>
  </v-row>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';

export default {
  name: 'ProductDetails',
  computed: {
    ...mapGetters(['product'])
  },
  mounted() {
    this.fetchProduct(this.$route.params.productId);
  },
  methods: {
    ...mapActions(['fetchProduct']),
    addToCart() {},
    removeFromCart() {}
  }
};
</script>

<style lang="scss" scoped></style>
