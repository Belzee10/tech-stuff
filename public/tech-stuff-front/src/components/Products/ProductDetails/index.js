export { default } from './ProductDetails.vue';
