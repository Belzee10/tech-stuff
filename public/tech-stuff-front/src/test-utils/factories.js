import { mount, createLocalVue } from '@vue/test-utils';
import mergeWith from 'lodash.mergewith';
import Vuetify from 'vuetify';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);

let vuetify;

beforeEach(() => {
  vuetify = new Vuetify();
});

const customizer = (_, srcValue) => {
  if (Array.isArray(srcValue)) {
    return srcValue;
  }
  if (srcValue instanceof Object && Object.keys(srcValue).length === 0) {
    return srcValue;
  }
};

/**
 * create component wrapper
 * @param {*} Component
 * @param {Object} overrides
 */
const createWrapper = (Component, overrides) => {
  const defaultMountingOptions = {
    localVue,
    vuetify
  };
  return mount(
    Component,
    mergeWith(defaultMountingOptions, overrides, customizer)
  );
};

/**
 * create Vuex store
 * @param {Object} overrides
 */
const createStore = overrides => {
  const defaultStoreConfig = {};
  return new Vuex.Store(mergeWith(defaultStoreConfig, overrides, customizer));
};

export { createWrapper, createStore };
