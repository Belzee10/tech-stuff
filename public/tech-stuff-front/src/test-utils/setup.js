import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.config.productionTip = false;

Vue.use(Vuetify);

/**
 * silent "console.error"
 */
global.console = {
  error: jest.fn(),
  log: console.log,
  warn: console.warn,
  info: console.info,
  debug: console.debug
};
