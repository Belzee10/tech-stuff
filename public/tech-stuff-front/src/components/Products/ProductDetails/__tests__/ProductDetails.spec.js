import { createWrapper, createStore } from '@/test-utils/factories';
import flushPromises from 'flush-promises';
import ProductDetails from '../ProductDetails.vue';

const productId = 1;

describe('ProductDetails.vue', () => {
  test('should dispatch a "fetchProduct" with the correct payload', async () => {
    const store = createStore();
    store.dispatch = jest.fn(() => Promise.resolve());

    createWrapper(ProductDetails, {
      store,
      mocks: {
        $route: {
          params: {
            productId
          }
        }
      }
    });
    await flushPromises();
    expect(store.dispatch).toHaveBeenCalledWith('fetchProduct', productId);
  });

  test('should render correctly', () => {
    const store = createStore();
    const wrapper = createWrapper(ProductDetails, {
      store,
      mocks: {
        $route: {
          params: {
            productId
          }
        }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
