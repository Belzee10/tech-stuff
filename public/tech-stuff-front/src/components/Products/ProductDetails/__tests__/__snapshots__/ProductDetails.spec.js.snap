// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails.vue should render correctly 1`] = `
<div
  class="row"
>
  <div
    class="col col-10 offset-1"
  >
    <div
      class="product-details v-card v-sheet theme--light"
    >
      <div
        class="container py-0 mt-10"
      >
        <div
          class="row"
        >
          <div
            class="col"
          >
            <span>
              Loading...
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
